vcap_utils/vcap_utils/backends/base_tensorrt.py

```diff
@@ -0,0 +1,278 @@
+        Args:
+            o1 (float): first argument of the result
+            o2 (float): second argument of the result
+            o3 (float): third argument of the result
+            o4 (float): fourth argument of the result
+            x: row index on the grid
+            y: column index on the grid
+
+        Returns:
+            float: rescaled first argument
+            float: rescaled second argument
+            float: rescaled third argument
+            float: rescaled fourth argument
+        """
+        o1 = (o1 - self.grid_centers_w[x]) * -self.box_norm
+        o2 = (o2 - self.grid_centers_h[y]) * -self.box_norm
+        o3 = (o3 + self.grid_centers_w[x]) * self.box_norm
+        o4 = (o4 + self.grid_centers_h[y]) * self.box_norm
+        return o1, o2, o3, o4
+
+    def parse_detection_results(
+            self, results: List[List[float]],
+            resize: Resize,
+            label_map: Dict[int, str],
+            min_confidence: float = 0.0,
+    ) -> List[DetectionNode]:
+        bbs = []
+        class_ids = []
+        scores = []
+        for c in label_map.keys():
+
+            x1_idx = c * 4 * self.grid_size
+            y1_idx = x1_idx + self.grid_size
+            x2_idx = y1_idx + self.grid_size
+            y2_idx = x2_idx + self.grid_size
+
+            boxes = results[0]
+            for h in range(self.grid_h):
+                for w in range(self.grid_w):
+                    i = w + h * self.grid_w
+                    score = results[1][c * self.grid_size + i]
+                    if score >= min_confidence:
+                        o1 = boxes[x1_idx + w + h * self.grid_w]
+                        o2 = boxes[y1_idx + w + h * self.grid_w]
+                        o3 = boxes[x2_idx + w + h * self.grid_w]
+                        o4 = boxes[y2_idx + w + h * self.grid_w]
+                        o1, o2, o3, o4 = self._apply_box_norm(o1, o2, o3, o4, w, h)
+                        xmin = int(o1)
+                        ymin = int(o2)
+                        xmax = int(o3)
+                        ymax = int(o4)
+                        bbs.append([xmin, ymin, xmax - xmin, ymax - ymin])
+                        class_ids.append(c)
+                        scores.append(float(score))
+        indexes = cv2.dnn.NMSBoxes(bbs, scores, min_confidence, 0.5)
+        detections = []
+        for idx in indexes:
+            idx = int(idx)
+            xmin, ymin, w, h = bbs[idx]
+            class_id = class_ids[idx]
+            class_name = label_map[class_id]
+            detections.append(
+                DetectionNode(
+                    name=class_name,
+                    coords=rect_to_coords(
+                        [xmin, ymin, (xmin + w), (ymin + h)]
+                    ),
+                    extra_data={"detection_confidence": scores[idx]},
+                )
+            )
+        resize.scale_and_offset_detection_nodes(detections)
+        return detections
```

Comment:
I don't understand how a lot of these random seeming calculations work. For example, each of these lines were a bit confusing:

```
x1_idx = class_id * 4 * self.grid_size

i = w + h * self.grid_w

score = results[1][class_id * self.grid_size + i]

o1 = boxes[x1_idx + w + h * self.grid_w]
o2 = boxes[y1_idx + w + h * self.grid_w]
o3 = boxes[x2_idx + w + h * self.grid_w]
o4 = boxes[y2_idx + w + h * self.grid_w]
```

If you could figure out how GridNet outputs these different things, it might be worth putting a docstring here explaining the output format and why this parsing works.

---
Also, please take a look at some of these proposed changes. I would prefer you take each change one at a time and test to make sure that there were no fundamental changes in how the function works. I haven't tested any of this. 

```suggestion
    def parse_detection_results(
            self, results: List[List[float]],
            resize: Resize,
            label_map: Dict[int, str],
            min_confidence: float = 0.0,
    ) -> List[DetectionNode]:
        nodes: List[DetectionNode] = []
        for class_id, class_name in label_map.items():
    
            x1_idx = class_id * 4 * self.grid_size
            y1_idx = x1_idx + self.grid_size
            x2_idx = y1_idx + self.grid_size
            y2_idx = x2_idx + self.grid_size
    
            boxes = results[0]
            for h in range(self.grid_h):
                for w in range(self.grid_w):
                    i = w + h * self.grid_w
                    score = results[1][class_id * self.grid_size + i]
                    if score < min_confidence:
                        continue
    
                    o1 = boxes[x1_idx + w + h * self.grid_w]
                    o2 = boxes[y1_idx + w + h * self.grid_w]
                    o3 = boxes[x2_idx + w + h * self.grid_w]
                    o4 = boxes[y2_idx + w + h * self.grid_w]
                    xmin, ymin, xmax, ymax = self._apply_box_norm(o1, o2, o3, o4, w, h)
                    nodes.append(
                        DetectionNode(
                            name=class_name,
                            coords=rect_to_coords([xmin, ymin, xmax, ymax]),
                            extra_data={"detection_confidence": float(score)},
                        )
                    )
    
        nodes = non_max_suppression(nodes, max_bbox_overlap=0.5)
        resize.scale_and_offset_detection_nodes(nodes)
```

Reply:
To be honest I don't understand how this postprocess code works. I found this code in this post.
https://forums.developer.nvidia.com/t/run-peoplenet-with-tensorrt/128000

It's sourced from this code:
https://github.com/Kwull/deepstream-4.0.1/blob/master/sources/libs/nvdsinfer_customparser/nvdsinfer_custombboxparser.cpp

Reply:
I think we need to take a closer look into the model itself to get a better understanding of the postprocessing.

And this code may not work for all detectors, I don't have enough samples to test with, at least it works for all detectors I found on TLT's website.

Reply:
Sure. For now, can you test my refactor? It's supposed to do exactly the same as what you originally had, but in fewer lines of code

Reply:
Yes I did, they are working as expected, and the performance is about the same